### PR TITLE
Mac user input detailed

### PR DESCRIPTION
**Pull request by royru**
Hi Chris,

This is the large PR i was mentioning yesterday. It features:

- added a second target (PersonalAnalytics Dev). Both run side-by-side with separate environment variables (see Environment.swift).
- detailed user input tracking similar to Windows
- fixes permissions for macOS 10.15 for Window name and URL tracking.
- added permission status overview to the Preference window (as shown yesterday)
- Resource Tracker: tracks FSEvents and resources associated with active application
- Added sparkle-project for auto-updates

Notice that I merged an older branch 'mac-environment' into this branch which was open for a long time. Therefore, commits from last year appear here even though they are already merged. New commits start on Jan 29, 2020. 

Let me know if I should walk you through all bits and pieces of this PR!

Cheers,
Roy

### CODE REVIEW

**Comment by csatterfield** [comment]
Hey @royru, I'm getting empty windows in the dev build for the retrospective and preferences after I granted accessibility and screen recording permissions & restarted. Not sure what's going on here...

image

**Comment by royru** [comment]
Hi Chris, thanks for the feedback! I will check the smaller comments/issues later this week. Regarding the empty windows: I did see something similar when I built the dev target for the first time after a while. Re-compiling 2-3 times helped. The problem suddenly disappeared and since then I haven't noticed this issue anymore. Very strange. Let me know if this issue persists for you!

File: src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift [hunk 1]
```diff
@@ -212,16 +206,16 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
...
     // MARK: - Setup, teardown of application including timers
     func applicationDidFinishLaunching(_ aNotification: Notification) {
-
-        if !AXIsProcessTrusted(){
-            launchPermissionPanel()
-            launchPermissionExplanationAlert()
-        }
+        
+        // opening the preference window to show which trackers are working and which are missing permission
+        showPreferences()
```
[the remark this answers is not in the document]
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift]: Updated this in the last commit with your proposal.

File: src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift [hunk 2]
```diff
@@ -70,41 +69,35 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
...
         pauseItem = NSMenuItem(title: "Pause Trackers", action: #selector(delegate.togglePause), keyEquivalent: "u")
+        let preferencesItem = NSMenuItem(title: "Preferences...", action: #selector(AppDelegate.showPreferences(_:)), keyEquivalent: "P")
+        let openDataItem = NSMenuItem(title: "Open Data Folder...", action: #selector(AppDelegate.openDataFolder(_:)), keyEquivalent: "O")
+        let checkUpdatesItem = NSMenuItem(title: "Check for Updates...", action: #selector(SUUpdater.checkForUpdates(_:)), keyEquivalent: "U")
+        // let toggleSummaryItem = NSMenuItem(title: "Toggle Survey", action: #selector(toggleSummary), keyEquivalent: "A")
+        // toggleSummaryItem.bind(NSBindingName(rawValue: "state"), to: defaultsController , withKeyPath: "values.\(AppConstants.summaryStateKey)", options: nil)
+        // let leaveReminderNotificationUntilClickedItem = NSMenuItem(title: "Notification Stays", action: Selector("togglePersistantUserNotifications:"), keyEquivalent: "P")
+        // leaveReminderNotificationUntilClickedItem.bind("state", to: defaultsController, withKeyPath: "values.\(AppConstants.notificationsPersistKey)", options: nil)
```
[the remark this answers is not in the document]
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift]: True. I moved it to `MenuBar.swift` and cleaned up the AppDelegate a bit (see last commit).

**Comment by csatterfield** [comment]
You're right, I'm using an older version of GRDB. I'll undo that change.

**Comment by csatterfield** [comment]
Empty windows problem seems to have been caused by lack of a project version. I added one in and I'm not seeing the issue anymore.

**Comment by csatterfield** [comment]
Thanks @royru, I'll merge this. Just make sure to add credit for the library you used 🙂

**Pull request merged**